### PR TITLE
Composing data lifecycle in index templates

### PR DESCRIPTION
**Pull request by gmarouli**
This PR implments the composing of the lifecycles of multiple component templates to a final lifecycle. The lifecycles are getting composed one level deep, meaning that the keys present on the latest lifecycle will override the ones of the others. If a key is missing then it keeps the value of the previous lifecycles. For example, in the following case we have two lifecycle both of which have the `data_retention` defined.
```
     [
       {
         "lifecycle": {
           "data_retention" : "10d"
         }
       },
       {
         "lifecycle": {
           "data_retention" : "20d"
         }
       }
     ]
```

The result will be `{ "lifecycle": { "data_retention" : "20d"}}`.

However, if we have the following two lifecycles:
```
     [
       {
         "lifecycle": {
           "data_retention" : "10d"
         }
       },
       {
       "lifecycle": { }
       }
     ]
```

The result will be `{ "lifecycle": { "data_retention" : "10d"} }` because the latest lifecycle does not the `data_retention` defined.

Part of: https://github.com/elastic/elasticsearch/issues/93596

### CODE REVIEW

**Comment by masseyke** [comment]
A few thoughts/questions:

We'll want good documentation on this because I'm guessing it'll be a frequent source of questions. Is documenting this blocked by #95326?

Right now there aren't any keys that conflict. For example I think we're fine if we have:
```
     [
       {
         "lifecycle": {
           "rollover" : "10d"
         }
       },
       {
         "lifecycle": {
           "data_retention" : "5d"
         }
       }
     ]
```
We'd just wind up with 
```
{ "lifecycle": { "rollover":"10d", "data_retention" : "5d"} }
```
But is there anything we plan on adding in the future where there _is_ some dependency between keys? I don't think there is, but thought I'd ask (because this would cause problems in that case).

If we have two of the same key, do we want to pick whichever one was last in the list vs giving the user an error that there is ambiguity? I assume the reasoning is that it's unavoidable -- sometimes you'll want to use two component templates that have a lifecycle defined? Is it worth reconsidering if we want to even support lifecycles in component templates for this reason? I'm just imagining future support cases where someone is caught by surprise.

**Comment by gmarouli** [comment]
Thank you for taking a look @masseyke.

> We'll want good documentation on this because I'm guessing it'll be a frequent source of questions. Is documenting this blocked by #95326?

This behavior is not documented yet but composing component templates is not new. You can see how the mapping one gets composed in [the index template doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-templates.html). I do agree though that we should demonstrate or at least reference this behavior when we have more comprehensive guides.
 
> Right now there aren't any keys that conflict. For example I think we're fine if we have:
> 
> ```
>      [
>        {
>          "lifecycle": {
>            "rollover" : "10d"
>          }
>        },
>        {
>          "lifecycle": {
>            "data_retention" : "5d"
>          }
>        }
>      ]
> ```
> 
> We'd just wind up with
> 
> ```
> { "lifecycle": { "rollover":"10d", "data_retention" : "5d"} }
> ```
> 
> But is there anything we plan on adding in the future where there _is_ some dependency between keys? I don't think there is, but thought I'd ask (because this would cause problems in that case).

Hm, I am not aware of such plans and I cannot think right now a good example of this, do you have something in mind? Without having something concrete in mind, I would expect that we can evaluate in two places:
- In the component template, if the data lifecycle configuration is invalid.
- In the index template, if the composed data lifecycle configuration is invalid.
Would that be sufficient, what do you think?

> If we have two of the same key, do we want to pick whichever one was last in the list vs giving the user an error that there is ambiguity? I assume the reasoning is that it's unavoidable -- sometimes you'll want to use two component templates that have a lifecycle defined? Is it worth reconsidering if we want to even support lifecycles in component templates for this reason? I'm just imagining future support cases where someone is caught by surprise.

I think we should keep the current behavior that the others already support. The component templates are ordered and each one can overwrite the previous one. Finally, the template on the index template itself overwrites everything else. Do you think that the lifecycle is different than the settings, mapping and aliases in a way that it poses a problem with this composing? 

I understand right now it might look a bit trivial, but with downsampling and other features that might come, I do think there is still value. For me it makes sense that it behaves similarly to settings etc.

**Comment by masseyke** [comment]
Thanks. That all makes sense. I just wanted to document that we'd thought about it. It's not different from anything else in a component template (it all seems like a source of potential mistakes, but we already support it).

**Review by masseyke — APPROVED** [review]
LGTM with the exception of the two minor unit test things I noted.

**Comment by gmarouli** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java [hunk 1]
```diff
@@ -1547,6 +1547,24 @@ public void testResolveLifecycle() throws Exception {
             // Based on precedence only the latest
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java:1562]: No, it is an omission, when reasoning about the tests, I kept mixing the lifecycles the high and low correspond to. That's why I changed to 1 and 2 :).

[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java]: Fixed

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java [hunk 2]
```diff
@@ -1547,6 +1547,24 @@ public void testResolveLifecycle() throws Exception {
...
+                1L,
+                null,
+                new ComposableIndexTemplate.DataStreamTemplate(),
+                null
+            );
+            state = service.addIndexTemplateV2(state, true, "my-template-4", it);
+
+            DataLifecycle resolvedLifecycle = MetadataIndexTemplateService.resolveLifecycle(state.metadata(), "my-template-4");
+
+            // Based on precedence only the latest
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java]: Fixed

**Pull request merged**